src/main/java/org/opensearch/plugin/insights/core/listener/QueryInsightsListener.java

```diff
@@ -82,31 +107,48 @@ public QueryInsightsListener(final ClusterService clusterService, final QueryIns
             this.queryInsightsService.validateWindowSize(type, clusterService.getClusterSettings().get(getTopNWindowSizeSetting(type)));
             this.queryInsightsService.setWindowSize(type, clusterService.getClusterSettings().get(getTopNWindowSizeSetting(type)));
         }
+
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(SEARCH_QUERY_METRICS_ENABLED_SETTING, v -> setSearchQueryMetricsEnabled(v));
+        setSearchQueryMetricsEnabled(clusterService.getClusterSettings().get(SEARCH_QUERY_METRICS_ENABLED_SETTING));
     }
 
     /**
-     * Enable or disable top queries insights collection for {@link MetricType}
+     * Enable or disable top queries insights collection for {@link MetricType}.
      * This function will enable or disable the corresponding listeners
      * and query insights services.
      *
      * @param metricType {@link MetricType}
      * @param isCurrentMetricEnabled boolean
      */
     public void setEnableTopQueries(final MetricType metricType, final boolean isCurrentMetricEnabled) {
-        boolean isTopNFeaturePreviouslyDisabled = !queryInsightsService.isTopNFeatureEnabled();
         this.queryInsightsService.enableCollection(metricType, isCurrentMetricEnabled);
-        boolean isTopNFeatureCurrentlyDisabled = !queryInsightsService.isTopNFeatureEnabled();
+        updateQueryInsightsState();
+    }
 
-        if (isTopNFeatureCurrentlyDisabled) {
-            super.setEnabled(false);
-            if (!isTopNFeaturePreviouslyDisabled) {
-                queryInsightsService.checkAndStopQueryInsights();
-            }
-        } else {
+    /**
+     * Set search query metrics enabled to enable collection of search query categorization metrics.
+     * @param searchQueryMetricsEnabled boolean flag
+     */
+    public void setSearchQueryMetricsEnabled(boolean searchQueryMetricsEnabled) {
+        this.queryInsightsService.enableSearchQueryMetricsFeature(searchQueryMetricsEnabled);
+        updateQueryInsightsState();
+    }
+
+    /**
+     * Update the query insights service state based on the enabled features.
+     * If any feature is enabled, it starts the service. If no features are enabled, it stops the service.
+     */
+    private void updateQueryInsightsState() {
+        boolean anyFeatureEnabled = queryInsightsService.isAnyFeatureEnabled();
+
+        if (anyFeatureEnabled && !super.isEnabled()) {
             super.setEnabled(true);
-            if (isTopNFeaturePreviouslyDisabled) {
-                queryInsightsService.checkAndRestartQueryInsights();
-            }
+            queryInsightsService.stop();
+            queryInsightsService.start();
+        } else if (!anyFeatureEnabled && super.isEnabled()) {
+            super.setEnabled(false);
+            queryInsightsService.stop();
```

Comment:
how is the concurrency handled on this stop/start? For example, if 2 features are enabled/disabled in single settings update, how the listeners for respective settings will consume those updated values. Are the listeners executed on cluster service thread?

Reply:
`AbstractScopedSettings` handles makes sure the `applySettings` method is synchronized